# Running Jupyter notebooks on the cloud

first steps to work with the notebooks in the cloud: Create and run a Microsoft Azure Notebooks! 
https://luisa0-luisacutillo78.notebooks.azure.com/
* Head to the website https://luisa0-luisacutillo78.notebooks.azure.com/ and click on sign-in (upper right corner). 
* If you do not have an account, click on create an account, use your email address and set a password.
* Make sure you are logged into your account.
* Go to the Notebook in this website that you want to run in the cloud and clone it
This should take you to the library where the notebooks are saved to.
Click on the notebook’s line (it should highlight the whole line) and "Shift+Enter" key combination will activate the commands.

# Our First Commands

In [22]:
3+9

[1] 12

In [23]:
3*9

[1] 27

## Help

In [24]:
help("median")



# Assignment operators and vectors


In [25]:
a <- 5
print('a=')
a
print('x=',x)
x
x <- c(1, 2, 3, 4, 5)
x

words <- c("eins", "zwei", "drei", "vier", "fuenf")
print('words=')
words
print(paste0('The first word is ',words[1]))
y <- c(1:100)
print('y=')
y

z <-seq(from=1, to=100, by=0.5)
print('z=')

z

w <- rep(c(1,2,3), 10)
print('w=')

w
print('length(w)=')
length(w)

[1] "a="


[1] 5

[1] "x="


NULL

[1] 1 2 3 4 5

[1] "words="


[1] "eins"  "zwei"  "drei"  "vier"  "fuenf"

[1] "The first word is eins"
[1] "y="


[1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100

[1] "z="


[1]   1.0   1.5   2.0   2.5   3.0   3.5   4.0   4.5   5.0   5.5   6.0   6.5
 [13]   7.0   7.5   8.0   8.5   9.0   9.5  10.0  10.5  11.0  11.5  12.0  12.5
 [25]  13.0  13.5  14.0  14.5  15.0  15.5  16.0  16.5  17.0  17.5  18.0  18.5
 [37]  19.0  19.5  20.0  20.5  21.0  21.5  22.0  22.5  23.0  23.5  24.0  24.5
 [49]  25.0  25.5  26.0  26.5  27.0  27.5  28.0  28.5  29.0  29.5  30.0  30.5
 [61]  31.0  31.5  32.0  32.5  33.0  33.5  34.0  34.5  35.0  35.5  36.0  36.5
 [73]  37.0  37.5  38.0  38.5  39.0  39.5  40.0  40.5  41.0  41.5  42.0  42.5
 [85]  43.0  43.5  44.0  44.5  45.0  45.5  46.0  46.5  47.0  47.5  48.0  48.5
 [97]  49.0  49.5  50.0  50.5  51.0  51.5  52.0  52.5  53.0  53.5  54.0  54.5
[109]  55.0  55.5  56.0  56.5  57.0  57.5  58.0  58.5  59.0  59.5  60.0  60.5
[121]  61.0  61.5  62.0  62.5  63.0  63.5  64.0  64.5  65.0  65.5  66.0  66.5
[133]  67.0  67.5  68.0  68.5  69.0  69.5  70.0  70.5  71.0  71.5  72.0  72.5
[145]  73.0  73.5  74.0  74.5  75.0  75.5  76.0  76.5  77.0  77.5  78.0  78.5
[157]  79.0  79.5  80.0  80.5  81.0  81.5  82.0  82.5  83.0  83.5  84.0  84.5
[169]  85.0  85.5  86.0  86.5  87.0  87.5  88.0  88.5  89.0  89.5  90.0  90.5
[181]  91.0  91.5  92.0  92.5  93.0  93.5  94.0  94.5  95.0  95.5  96.0  96.5
[193]  97.0  97.5  98.0  98.5  99.0  99.5 100.0

[1] "w="


[1] 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3 1 2 3

[1] "length(w)="


[1] 30

In [32]:
z <-seq(from=1, to=10, by=0.2)
print('z=')
z

[1] "z="


[1]  1.0  1.2  1.4  1.6  1.8  2.0  2.2  2.4  2.6  2.8  3.0  3.2  3.4  3.6  3.8
[16]  4.0  4.2  4.4  4.6  4.8  5.0  5.2  5.4  5.6  5.8  6.0  6.2  6.4  6.6  6.8
[31]  7.0  7.2  7.4  7.6  7.8  8.0  8.2  8.4  8.6  8.8  9.0  9.2  9.4  9.6  9.8
[46] 10.0

In [28]:
sprintf("Current working dir: %s", getwd())

[1] "Current working dir: /home/nbuser/library"

## Index Operators and Vector Operations

In [24]:
#############################
# Index operator: [...]     #
#############################
w[1:4]
w[seq(from=2, to=17, by=3)]



########################
# Vector operations    #
########################
x<-c(1,2,3)
y<-c(2,4,6)

x + y
10*x
x*y

x < 3



[1] 1 2 3 1

[1] 2 2 2 2 2 2

[1] 3 6 9

[1] 10 20 30

[1]  2  8 18

[1]  TRUE  TRUE FALSE

In [34]:
x=c(1,2,4)
x

[1] 1 2 4

In [36]:
y=c(1,1,1)
y

[1] 1 1 1

In [37]:
x+y

[1] 2 3 5

### Programming constructs 

In [25]:

# Branching with "if / else"
if(x[2] < -4) {
 r <- 1
} else
{
 r <- 0
} 
r

#for loops
for (i in c(1:3)){
print(x[i])
}

for (i in c(1:5)){
print(words[i])
}


[1] 0

[1] 1
[1] 2
[1] 3
[1] "eins"
[1] "zwei"
[1] "drei"
[1] "vier"
[1] "fuenf"


In [39]:
x=c(1,5,6)
x[2]

[1] 5

In [40]:
# Branching with "if / else"
if(x[2] < -4) {
 r <- 1
} else
{
 r <- 0
} 
r

[1] 0

In [41]:
#for loops
for (i in c(1:3)){
print(x[i])
}


[1] 1
[1] 5
[1] 6


## Generic mathematical Functions

In [ ]:
exp(5)
log(3)
log(exp(10))

a<-8
sqrt(a)


## Generate random numbers  


In [30]:
####################################  
# Pseudo-random sampling: sample() #
####################################
sample(x, 20, replace=FALSE)      #without replacement
sort(sample(x, 20, replace=TRUE)) #with replacement
x_part <- x[1:10]

# Use your own sampling probabilities:
prob_x <- c(0.004, 0.9, 0.3, 0.001, 0.6, 0.005, 0.12, 0.1, 0.0006, 1)
x_part
sample(x_part, replace=TRUE, prob=prob_x)

# 100 Bernoulli trials
sample(c(0,1), 100, replace = TRUE)




[1]  1.81769804  0.11137547 -1.49224337  1.16330771 -0.37156871  1.19063391
 [7] -0.15282843  0.83608333  1.78026899  2.00013620  2.48656687  0.38473890
[13] -1.15835795  0.91684017 -2.05884386  0.59532870  1.74892341  0.48593978
[19]  0.56912223  0.02817881

[1] -2.1967083 -2.0588439 -1.9978058 -1.5627437 -1.4540716 -1.0229945
 [7] -0.7701250 -0.5457350 -0.5051582 -0.4902495 -0.4235603 -0.4199242
[13] -0.2140177  0.6723688  0.9168402  1.1184961  1.1184961  1.7802690
[19]  2.0001362  2.0001362

[1]  2.0001362  0.6145009 -0.8796301  0.3847389 -0.3715687  3.0802585
 [7]  2.4865669 -1.7480138  1.1926995 -1.0414728

[1]  0.6145009  0.6145009 -1.0414728 -0.3715687  0.6145009 -0.3715687
 [7]  0.6145009 -0.8796301 -0.8796301 -1.0414728

[1] 0 0 0 0 1 0 0 1 0 1 1 1 0 1 1 0 0 0 0 0 1 0 1 1 0 1 1 1 0 1 1 0 1 0 0 0 1
 [38] 1 0 1 1 1 1 1 0 0 1 0 0 0 1 0 1 0 1 1 1 1 1 0 1 0 0 1 0 1 1 0 1 1 0 1 1 1
 [75] 0 0 1 0 1 1 1 0 0 0 1 0 0 1 0 0 1 0 1 1 0 0 0 1 1 1

In [42]:
sample(c(0,1), 100, replace = TRUE)


[1] 0 1 1 1 0 1 1 1 1 1 0 1 1 0 1 1 0 1 0 1 0 1 0 0 0 0 1 0 1 1 0 1 0 0 1 1 0
 [38] 1 1 1 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 0 1 0 0 1 0 0 1 0 1 0 0 0 1 1 1 1 0
 [75] 1 0 1 1 0 1 0 1 0 0 1 1 0 1 0 0 1 0 1 0 0 0 0 0 1 0

# Create a user defined function in R

# # One of the great strengths of R is the user's ability to add functions. 
## In fact, many of the functions in R are actually functions of functions.
## The structure of a function is given below.



### <span style="color:blue"> your_function <- function(arg1, arg2, ... ){ </span>

### <span style="color:blue"> statements</span>
### <span style="color:blue">  return(object)</span>
### <span style="color:blue"> } </span>

### **EXAMPLE** Define a function performing the sum of squares

In [43]:
sum.of.squares <- function(x,y) {
  z=x^2 + y^2
    return(z)
    
}


In [46]:
# how to call it
x=1; y=2; 
z=sum.of.squares(x,y)
print(z)

[1] 5


# More in general, to load your function or your R code stored in "namefile.R", use the command source("namefile.R")
  

##  <span style="color:red"> Exercise: Impement your own function to compute the square of a binomial (x+y)^2

In [48]:
binomsquare= function(x,y){
    z=x^2+y^2+2*x*y
    return(z)
    
}

In [50]:
x=1;y=2;
m=binomsquare(x,y)
print(m)

[1] 9


# How to find the factorial of an integer number

In [52]:
# take input from the user
num = as.integer(readline(prompt="Enter a number: "))
factorial = 1
# check is the number is negative, positive or zero
if(num < 0) {
print("Sorry, factorial does not exist for negative numbers")
} else if(num == 0) {
print("The factorial of 0 is 1")
} else {
for(i in 1:num) {
factorial = factorial * i
}
print(paste("The factorial of", num ,"is",factorial))
}

Enter a number: 8
[1] "The factorial of 8 is 40320"


## Here, we take input from the user and check if the number is negative, zero or positive using if...else statement.

## If the number is positive, we use for loop to calculate the factorial.

## We can also use the built-in function factorial() for this.


In [53]:
factorial(8)

[1] 40320

## *NB* be carefulf with the factorial fuction! Calculators have memoy limits!

If you ever heard the expression of 32-bit computers or 64-bit computers, then you are already pretty close to figuring out how large a number can be.

You probably know that computers calculate in binary (we humans commonly do in decimals). A bit is what we could call a digit. Now, the largest number you can achieve with one digit is 9. And with two digits it's 99. And so on.

In binary, you usually have units with size either 32 bits or 64 bits respectively.
I suggest you to learn how binary works but I try to make a quick analogy:

67 = 10^1 * 6 + 10^0 * 7 = 10 * 6 + 1 * 7 = 60 + 7
121 = 10^2 * 1 + 10^1 * 2 + 10^0 + 1 = 100 + 20 + 1
So each digit is related with a power of ten. In binary, it's the same but with powers of two! Also, in binary, we only have the numbers 1 and 0. Thus:

10 = 2^1 * 1 + 2^0 * 0 = 2 * 1 + 1 * 0 = 2 + 0 = 2
1011 = 2^3 * 1 + 2^2 * 0 + 2^1 * 1 + 2^0 * 1 = 8 * 1 + 0 + 2 + 2 * 1 = 12
Now, assuming that we leave out negative numbers, with 32 bits, the largest numbers is 32 ones:

11111111111111111111111111111111 = 4,294,967,295


# Birthday Problem


In a classroom of $n$ students, what is the probability $p_n$ that two (or more) students share
the same birthday?

It's easier to compute the probability that no two students share a birthday.
Let's look at the students one at a time.
The first student can have any birthday he/she likes.
The second student can not share the first student's birthday: 364 choices.
The third student cannot share either of the first two birthdays: 363 choices.
... Etc. ...
The $n-th$ student cannot share any of the previous $n - 1$ birthdays: $365 - n + 1$ choices.

Therefore, probability of no shared birthdays is
\begin{equation}
\frac{365\times364 \times \ldots (365 - n + 1)}{365^n}
\end{equation}

and the probability of a shared birthday is
\begin{equation}
1 - \frac{365\times364 \times \ldots (365 - n + 1)}{365^n}
\end{equation}

For $n = 23$, the answer is $p_{23} = 0.507$

In [54]:
birthday <- function(n){
  p = 1;
  for(a in seq(365,365-n+1,-1)){
    p = p * a/365;
    }
  return(1-p)
  }



In [55]:
birthday(23)

[1] 0.5072972

## R built in function: pbirthday()

In [56]:
pbirthday(23)

[1] 0.5072972